ouroboros-consensus/src/Ouroboros/Consensus/Mempool/API.hs

```diff
@@ -85,13 +85,14 @@ data Mempool m blk idx = Mempool {
       -- persistence.
       --
       -- The transactions will be validated, /in order/, before they are added
-      -- to the mempool. Invalid transactions will be returned along with the
-      -- validation error. In principle it is possible that such validation
-      -- errors are transient; for example, it is possible that a transaction is
-      -- rejected because one of its inputs is not /yet/ available in the UTxO
-      -- (the transaction it depends on is not yet in the chain, nor in the
-      -- mempool). In practice however it is likely that rejected transactions
-      -- will still be rejected later, and should just be dropped.
+      -- to the mempool. Newly added valid transactions, as well as all invalid
```

Comment:
You should be a bit more precise in how the output relates to the input. Will they have the same length? Does "all invalid transactions" mean *the invalid transactions among the new transactions given to `addTxs`* or *all the invalid transactions that are now in the mempool, including the new transactions given to `addTxs`*?

You could formulate a simple property here, e.g., `sort (map snd input) === sort (map fst output)` (if there is an `Ord` instance for `GenTx`).

Reply:
Ah, so it is the second (including  those already in the mempool), as stated in the PR description:
> addTxs should return both valid transactions which have been newly added to the Mempool, as well as all invalid transactions not added to or removed from the Mempool.

This is not what a user would typically expect based on the type signature (that's why we have a docstring of course). Hmm, let me try to remember why we include invalid transactions that were already in the Mempool (correct me where I'm wrong): after adding new transactions, existing valid transactions could become invalid (input already spent or the transaction is already in the chain), and by returning these here, we can trace them and notify the submitter?

Could existing invalid transactions also become valid (a missing input is made available by a new transaction)?

When do we plan on removing invalid transactions from the Mempool again? Will this only be based on "age"?

Reply:
@mrBliss Yeah, that's correct.

For example, as you noted, when you attempt add new transactions, all of the existing transactions in the `Mempool` are validated with respect to the _current_ ledger state of the `ChainDB` (at that particular time). Things could have changed since the last call to `addTxs` or `syncState` so some of these transactions, which were once valid at some point in time, could be rendered invalid and removed from the mempool. Likely for reasons such as those you mentioned, i.e. input already spent or transaction already included in a block.

> Could existing invalid transactions also become valid (a missing input is made available by a new transaction)?

Hmm, I'm interpreting this as: a once valid transaction in the mempool becomes invalid due to one of its referenced inputs being utilized by another transaction on the chain. I'm not sure whether this is possible. I don't think that a once valid, now invalid, transaction in the mempool could ever become valid again. In order for that transaction to become valid again, I think you'd have to somehow introduce a valid transaction into the chain with the exact same hash as the one which was previously referenced as well as a valid output at the specified index (since a `TxIn` is a reference to a `TxId` and an index of an output of that referenced transaction).

Actually, maybe it's possible for a currently invalid transaction to become valid again in the event of a rollback?

Oh, on the other hand, if what you meant was something like: 2 transactions, one of which is dependent upon an output of the other, are submitted to the mempool. However, they're either sent or received in the wrong order and, as a result, the transaction which is depend on the other, fails validation. In this case, I think it's entirely possible for a once invalid transaction to become valid in the future. E.g., in this case, the invalid transaction would potentially become valid after the other is accepted into the mempool.

> When do we plan on removing invalid transactions from the Mempool again? Will this only be based on "age"?

Transactions which are found to be invalid should be removed from the `Mempool` upon any call to `addTxs` or `syncState`. There's also the expiry functionality mentioned in issue #622 (which I think you're referring to). I don't think we decided on an actual time period in which a transaction would expire but I think we did consider measuring this time in slots. We would probably remove expired transactions upon calls to `syncState` (which is called whenever the `ChainDB` "tip point" changes in a thread spawned by the function `forkSyncStateOnTipPointChange `).

Reply:
Since this seems a bit odd and, perhaps, counter-intuitive given the type signature, should we come up with another idea as to what `addTxs` should return? I'd be a bit skeptical, however, of just silently dropping the invalid transactions which were previously in the mempool upon a call to `addTxs`. 

Maybe I should try to make the documentation more clear?

Reply:
Maybe a more specific `ApplyTxErr` for each case? Although, this burden would be placed upon the specific `ApplyTx` implementation as `ApplyTxErr` is defined as such:
```
class UpdateLedger blk => ApplyTx blk where
  ...
  type family ApplyTxErr blk :: *
  ...
```

Reply:
> Oh, on the other hand, if what you meant was something like: 2 transactions are submitted to the mempool, one of which is dependent upon an output of the other. However, they're either sent or received in the wrong order and, as a result, the transaction which is depend on the other, fails validation. In this case, I think it's entirely possible for a once invalid transaction to become valid in the future. E.g., in this case, the invalid transaction would potentially become valid after the other is accepted into the mempool.

That's what I meant :slightly_smiling_face: 

> > When do we plan on removing invalid transactions from the Mempool again? Will this only be based on "age"?
> 
> Transactions which are found to be invalid should be removed from the `Mempool` upon any call to `addTxs` or `syncState`. There's also the expiry functionality mentioned in issue #622 (which I think you're referring to). I don't think we decided on an actual time period in which a transaction would expire but I think we did consider measuring this time in slots. We would probably remove expired transactions upon calls to `syncState` (which is called whenever the `ChainDB` "tip point" changes in a thread spawned by the function `forkSyncStateOnTipPointChange `).

Ok. I suppose that not every caller of `addTxs` *has to* remove the invalid transactions it returns. I think after `syncState` is a more logical point to remove them.

> Since addTxs seems a bit odd and, perhaps, counter-intuitive given the type signature, should we come up with another idea as to what addTxs should return? I'd be a bit skeptical, however, of just silently dropping the invalid transactions which were previously in the mempool.
>
> Maybe I should try to make the documentation more clear?

Yes, keep the current behaviour, but make the documentation more clear, e.g., by explaining why we're doing it this way, and how valid transactions can become invalid and vice versa, etc. Also make sure it mentions that when you add an invalid transaction, it *will* be stored in the mempool, because it might become valid later (sounds like another QC property :slightly_smiling_face:).

> Maybe a more specific ApplyTxErr for each case?

No, I don't think that's needed.

Reply:
> Also make sure it mentions that when you add an invalid transaction, it will be stored in the mempool, because it might become valid later

Oh, I should clarify (perhaps my explanation was a bit confusing): when a transaction is found to be invalid, it is _not_ stored in the `Mempool`.

Here's what Edsko noted about why we drop invalid transactions in the `Mempool` (see the doc for `addTxs`):
> In principle it is possible that such validation errors are transient; for example, it is possible that a transaction is rejected because one of its inputs is not /yet/ available in the UTxO (the transaction it depends on is not yet in the chain, nor in the mempool). In practice however it is likely that rejected transactions will still be rejected later, and should just be dropped.

Reply:
Ah, I see. Make sure that the documentation mentions this so that other people won't get confused like me :slightly_smiling_face: